R/add_correlated_data.R

```diff
@@ -457,24 +511,41 @@ addCorGen <- function(dtOld, nvars=NULL, idvar = "id", rho=NULL, corstr=NULL, co
       
       # check if the dimensions of corr matrix matches (equal) cluster size
       
-      dn <- dtTemp[, .N, keyby = .id]
-      dn[, dim := nrow(corMatrix)]
-      compare_cluster_size <- dn[, sum(N != dim)]
-      if (compare_cluster_size != 0) {
-        stop("Dimensions of corMatrix not equal to cluster sizes!")
+      if (!wide) {
+        dn <- dtTemp[, .N, keyby = .id]
+        dn[, dim := nrow(corMatrix)]
+        compare_cluster_size <- dn[, sum(N != dim)]
+        if (compare_cluster_size != 0) {
+          stop("Dimensions of corMatrix not equal to cluster sizes!")
+        }
       }
     }
   }
   
-  
   if (wide) { # Convert to long form temporarily
+    if ( is.null(nvars) ) nvars <- nrow(corMatrix)
     dtTemp <- addPeriods(dtTemp, nPeriods = nvars, idvars = ".id")
   }
   
   dtTemp[, seq_ := 1:.N, keyby = .id]
-  nvars <- dtTemp[.id == 1, .N] # only permits case where number of records per id is the same
-
-  ####
+  # nvars <- dtTemp[.id == 1, .N] # only permits case where number of records per id is the same
+  
+  counts <- dtTemp[, .N, by = .id][, N]
+  same_nvar <- all(counts == counts[1])
+  
+  if (!wide) {    # multiple record per id
+    if (is.null(corMatrix)) {
+      if (same_nvar) {
```

Comment:
It's fine here imo because it's just the one line each but when you nest three levels deep it's usually a good time to check if pulling some of the logic out can flatten things out a bit. (a goto example: instead of wrapping the entire function with `if(condition)` invert the condition and return early removing that level of nesting from the entire function)